### PR TITLE
[saas-file-owners] check that diff is in changed paths

### PR DESCRIPTION
**Pull request by maorfr**
related to https://issues.redhat.com/browse/APPSRE-3029

replaces #1385

we are seeing issues of incorrect comments on MRs, where the saas-file-owners integration believes it is changing saas files which it does not. I've investigated a bit and my current working assumption is that this happens when a merge was made to master but the build-master job has not yet completed or has completed and the data was not yet reloaded in the server. This causes a MR to believe it is the one changing the other saas file.

### CODE REVIEW

**Review by apahim — APPROVED** [review]
Makes sense... graphql and the git repo are not guaranteed to be always in sync.

**Comment by erdii** [comment]
lgtm!

File: reconcile/saas_file_owners.py [hunk 1]
```diff
@@ -101,11 +102,24 @@ def collect_baseline():
...
+def collect_compare_diffs(current_state, desired_state, changed_paths):
     """ Collect a list of URLs in a git diff format
     for each change in the merge request """
     compare_diffs = set()
     for d in desired_state:
+        # check if this diff was actually changed in the current MR
+        changed_path_matches = [c for c in changed_paths
+                                if c.endswith(d['saas_file_path'])]
+        if not changed_path_matches:
+            # this diff was found in the graphql endpoint comparisson
```
Comment (rporres) [inline reconcile/saas_file_owners.py:114]: s/comparisson/comparisson/

File: reconcile/saas_file_owners.py [hunk 2]
```diff
@@ -101,11 +102,24 @@ def collect_baseline():
...
+            # but is not a part of the changed paths.
+            # the only knows case for this currently is if a previous MR
```
Comment (rporres) [inline reconcile/saas_file_owners.py:116]: s/knows/known/

File: reconcile/saas_file_owners.py [hunk 3]
```diff
@@ -101,11 +102,24 @@ def collect_baseline():
...
+            # that chages another saas file was merged but is not yet
```
Comment (rporres) [inline reconcile/saas_file_owners.py:117]: s/chages/changes/

File: reconcile/saas_file_owners.py [hunk 4]
```diff
@@ -290,6 +305,20 @@ def run(dry_run, gitlab_project_id=None, gitlab_merge_request_id=None,
...
+        saas_file_path = diff['saas_file_path']
+        changed_path_matches = [c for c in changed_paths
+                                if c.endswith(saas_file_path)]
+        if not changed_path_matches:
+            # this diff was found in the graphql endpoint comparisson
+            # but is not a part of the changed paths.
+            # the only knows case for this currently is if a previous MR
+            # that chages another saas file was merged but is not yet
+            # reflected in the baseline graphql endpoint.
+            # https://issues.redhat.com/browse/APPSRE-3029
```
Comment (rporres) [inline reconcile/saas_file_owners.py:318]: see above typos proposed fixes

**Comment by rporres** [comment]
>Makes sense... graphql and the git repo are not guaranteed to be always in sync.

But we could make that happen as we have the commit info in the bundle...

File: reconcile/saas_file_owners.py (hunk 1 above)
Comment (maorfr) [reply reconcile/saas_file_owners.py:114]: comparison? :)

Comment (maorfr) [reply reconcile/saas_file_owners.py:114]: https://github.com/app-sre/qontract-reconcile/pull/1408

File: reconcile/saas_file_owners.py (hunk 4 above)
Comment (maorfr) [reply reconcile/saas_file_owners.py:318]: https://github.com/app-sre/qontract-reconcile/pull/1408

Comment (rporres) [reply reconcile/saas_file_owners.py:318]: thx!

File: reconcile/saas_file_owners.py (hunk 1 above)
Comment (rporres) [reply reconcile/saas_file_owners.py:114]: LOL. yes!

**Pull request merged**